content/blog/_posts/2021-05-java-client.md

```diff
@@ -0,0 +1,660 @@
+
+Simply download [Adam Heinrich's NativUtils.java](https://raw.githubusercontent.com/adamheinrich/native-utils/master/src/main/java/cz/adamh/utils/NativeUtils.java) to
+`lib/src/main/java/my/java/lib/NativeUtils.java`, change the package to
+`my.java.lib`.
+
+Finally, we go back to your flaipgen `Foo` class and update it as follows:
+
+```rust
+foreign_class!(class Foo {
+    self_type Foo;
+    constructor Foo::new(_: i32) -> Foo;
+    fn Foo::set_field(&mut self, _: i32);
+    fn Foo::val(&self) -> i32;
+    foreign_code r#"
+        static {
+            try {
+                NativeUtils.loadLibraryFromJar("/libmy_java_lib.so");
+            } catch (java.io.IOException e) {
+                e.printStackTrace();
+            }
+        }"#;
+});
+```
+
+The `foreign_code` block will add a `static { ... }` routine to the `Foo` class
+declaration in java.
+
+## Testing it all out
+
+In our [FooTest.java](#footestjava), we wrote an assert function that would
+fail. Let's verify that this actually fails. Do this by running `./gradlew test`. It should fail with the following assertion error:
+
+```bash
+$ ./gradlew test
+
+> Task :lib:cargo-build
+    Finished release [optimized] target(s) in 0.04s
+
+> Task :lib:test FAILED
+JNI_OnLoad begin
+
+my.java.lib.FooTest > testSomeLibraryMethod FAILED
+    java.lang.AssertionError at FooTest.java:10
+
+2 tests completed, 1 failed
+
+FAILURE: Build failed with an exception.
+
+* What went wrong:
+Execution failed for task ':lib:test'.
+> There were failing tests. See the report at: file:///home/simlay/projects/infinyon/fluvio-demo-apps-rust/my-java-lib-blog-post/lib/build/reports/tests/test/index.html
+
+* Try:
+Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.
+
+* Get more help at https://help.gradle.org
+
+BUILD FAILED in 943ms
+6 actionable tasks: 2 executed, 4 up-to-date
+```
+
+Now, let's update our `FooTest.java` to the following:
+
+```java
+package my.java.lib;
+
+import org.junit.Test;
+import static org.junit.Assert.*;
+import my.java.lib.Foo;
+
+public class FooTest {
+    @Test public void testSomeLibraryMethod() {
+        Foo foo = new Foo(10);
+        assertTrue("Foo.val", foo.val() == 10);
+        foo.set_field(15);
+        assertTrue("Foo.val", foo.val() == 15);
+    }
+}
+```
+
+And you should see:
+
+```bash
+$ ./gradlew test
+
+> Task :lib:cargo-build
+    Finished release [optimized] target(s) in 0.04s
+
+    BUILD SUCCESSFUL in 624ms
+    6 actionable tasks: 1 executed, 5 up-to-date
+```
+
+And there you go! You've now called rust from java and can distribute your rust code
+in a java jar! :tada:
+
+## Conclusion
+
+This post is a bit longer and the code is a bit more verbose than I normally
+post about but this is a pretty clean setup. You can get the source for this
+post in our [fluvio-demo-apps-rust] repository.
```

Comment:
Is this right? Why `fluvio-demo-apps-rust`?

Reply:
Yeah, I'm not sure where to put the code. We might wanna move to a `blog-post-code` repo or something.

Reply:
Oh ok I missed the fact you had a PR up there, I just thought you pasted the wrong repo or something. In that case I'm good with it 👍